docs/src/variables.md

```diff
@@ -580,6 +587,57 @@ julia> x = @variable(model, [i=1:2], base_name="x", lower_bound=i, integer=true)
     Creating two named JuMP variables with the same name results in an error at
     runtime. Use anonymous variables as an alternative.
 
+## Constrained variables
+
+To create variables constrained to belong to the [`SecondOrderCone`](@ref) either
+of the following two methods can be used:
+```jldoctest constrained_variables; setup=:(model=Model())
+julia> @variable(model, x[1:3] in SecondOrderCone())
+3-element Array{VariableRef,1}:
+ x[1]
+ x[2]
+ x[3]
+
+julia> @variable(model, y[1:3])
+3-element Array{VariableRef,1}:
+ y[1]
+ y[2]
+ y[3]
+
+julia> @constraint(model, y in SecondOrderCone())
+[y[1], y[2], y[3]] ∈ MathOptInterface.SecondOrderCone(3)
+```
+The variables `x` in the example above are called *constrained variables* in
+contrast which the variables `y` which are called *free variables* because
+the constraint added on the variables is added after their creation as free
+variables.
+
+!!! note
+    Variable bounds and integrality constraints are specified when variables
+    are created, but they don't involve constrained variables, unless you
+    explicitly use `in` or the `set` keyword. Therefore `@variable(model, z <= 1,
+    Int)` and `@variable(model, upper_bound = 1, integer = true)` are not
+    constrained variables but `@variable(model, z in MOI.LessThan(1.0), Int)` is
+    a constrained variable in `MOI.LessThan(1.0)` and `@variable(model,
+    lower_bound = 1.0, set = MOI.Integer())` is a constrained variable in
+    `MOI.Integer()`.
```

Comment:
I think the name "constrained variables" is deceivingly simple for the concept it represents. Could we open this for bikeshedding? What about "variables constrained on creation"?

Reply:
Yes, let's discuss the name. "variable added constrained" is another possibility

Reply:
How about

---

## Restricted variables

By default, `@variable(model, x)` creates a _free_ variable that belongs to the 
set of real numbers.

If you add constraints, such as `@variable(model, x >= 0, Int)`, then the 
`@variable` macro is equivalent to:
```julia
@variable(model, x)
@constraint(model, x in MOI.GreaterThan(0.0))
@constraint(model, x in MOI.Integer())
```
Importantly, the bound and integrality constraints are added _after_ the 
variable has been created.

However, in some cases, it is necessary to supply a constraining set _at 
creation time_. We call these variables _restricted_ variables to differentiate 
them from _free_ variables.

For example, one way to create a vector of variables constrained to the second 
order cone is as follows:
```jldoctest constrained_variables; setup=:(model=Model())
julia> @variable(model, x[1:3])
3-element Array{VariableRef,1}:
x[1]
x[2]
x[3]

julia> @constraint(model, x in SecondOrderCone())
[x[1], x[2], x[3]] ∈ MathOptInterface.SecondOrderCone(3)
```

An alternative approach, creating a vector of variables _restricted_ to the 
[`SecondOrderCone`](@ref) is:
```jldoctest constrained_variables
julia> @variable(model, y[1:3] in SecondOrderCone())
3-element Array{VariableRef,1}:
y[1]
y[2]
y[3]
```
Importantly, _the second-order cone restriction on `y` cannot be deleted.

!!! info
    Restricted variables are most helpful for creating PSD matrices, where 
    creating a matrix of variables _restricted_ to the PSD cone can result in a
    smaller problem than _constraining_ the matrix to the PSD cone after 
    variable creation. Therefore, prefer
    ```julia
    @variable(model, x[1:2, 1:2] in PSDCone())
    ```
    over
    ```julia
    @variable(model, x[1:2, 1:2], PSD)
    ```

### The `set` keyword

An alternate syntax to `x in Set` is to use the `set` keyword of 
[`@variable`](@ref). This is most useful when creating anonymous variables:
```julia
x = @variable(model, [1:2, 1:2], set = PSDCone())
```

Reply:
Changing "constrained" to "restricted" doesn't really resolve my concern. Mathematically, variables can be restricted by adding constraints after they're created. Also, if you ask anyone doing linear programming if `@variable(model, x >= 0)` creates a free variable, they'll guess no. 

What we're *actually* trying to distinguish is constraints specified at creation time, so why not put the creation-time concept in the name? Deceptively short names create confusion.

We can have "variables constrained on creation" and "variables free on creation". "variable added constrained" doesn't quite parse as well for me.

Reply:
> Changing "constrained" to "restricted" doesn't really resolve my concern. 

I guess I was aiming at "constrained" being an overloaded term, and "restricted" has a similar meaning but we don't use it presently.

> We can have "variables constrained on creation" and "variables free on creation". 

I'm good with this.

Reply:
I have updated the PR with the new name and improved the doc with @odow's suggestions